### PR TITLE
UCP/PROTO/TAG: Add eager zero-copy, short, and bcopy/single protocols

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/ucp/proto/proto_multi.inl [hunk 1]
```diff
@@ -108,4 +108,30 @@ ucp_proto_multi_progress(ucp_request_t *req, ucp_proto_send_multi_cb_t send_func
...
+    ucp_request_t *req                 = ucs_container_of(uct_req, ucp_request_t,
+                                                          send.uct);
+    const ucp_proto_multi_priv_t *priv = req->send.proto_config->priv;
+    ucs_status_t status;
+
+    if (!(req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED)) {
+        status = ucp_proto_request_zcopy_init(req, priv->reg_md_map, comp_func);
+        if (status != UCS_OK) {
+            ucp_proto_request_zcopy_complete(req, status);
+            return UCS_OK;
```
Comment (dmitrygx) [inline src/ucp/proto/proto_multi.inl:125]: [q] why `UCS_OK`?

File: src/ucp/proto/proto_single.inl [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                               ucp_lane_index_t lane,
+                               uct_pack_callback_t pack_func, void *pack_arg,
+                               size_t max_packed_size)
+{
+    ucp_ep_t *ep                       = req->send.ep;
+    const uct_iface_attr_t *iface_attr = ucp_ep_get_iface_attr(ep, lane);
+    ssize_t packed_size;
+    uint64_t *buffer;
+
+    if ((max_packed_size <= UCS_ALLOCA_MAX_SIZE) &&
```
Comment (dmitrygx) [inline src/ucp/proto/proto_single.inl:27]: `max_packed_size <= ucs_min(UCS_ALLOCA_MAX_SIZE, iface_attr->cap.am.max_short)`?

File: src/ucp/proto/proto_single.inl [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    ucs_assert(req->send.dt_iter.offset == 0);
+
+    if (!(req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED)) {
+        md_map = (spriv->reg_md == UCP_NULL_RESOURCE) ? 0 : UCS_BIT(spriv->reg_md);
+        status = ucp_proto_request_zcopy_init(req, md_map,
+                                              ucp_proto_request_zcopy_completion);
+        if (status != UCS_OK) {
+            ucp_proto_request_zcopy_complete(req, status);
+            return UCS_OK;
```
Comment (dmitrygx) [inline src/ucp/proto/proto_single.inl:95]: [q] why UCS_OK?

File: src/ucp/tag/eager_single.c [hunk 4]
```diff
@@ -8,13 +8,40 @@
...
+                             UCP_AM_ID_EAGER_ONLY, req->send.msg_proto.tag.tag,
+                             req->send.dt_iter.type.contig.buffer,
+                             req->send.dt_iter.length);
+    if (ucs_unlikely(status == UCS_ERR_NO_RESOURCE)) {
+        req->send.lane = spriv->super.lane; /* for pending add */
+        return status;
+    }
+
+    /* UCS_INPROGRESS is not expected */
+    ucs_assert((status == UCS_OK) || UCS_STATUS_IS_ERR(status));
```
Comment (dmitrygx) [inline src/ucp/tag/eager_single.c:39]: maybe just `status != UCS_INPROGRESS`

File: src/ucp/proto/proto_single.inl [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                                   ucp_proto_complete_cb_t error_func)
+{
+    ucs_status_t status;
+
+    ucs_assert(error_func != NULL);
+
+    status = ucp_proto_am_bcopy_single_send(req, am_id, lane, pack_func,
+                                            pack_arg, max_packed_size);
+    if (ucs_likely(status == UCS_OK)) {
+        if (complete_func != NULL) {
```
Comment (dmitrygx) [inline src/ucp/proto/proto_single.inl:62]: it is always not a `NULL` (in the `ucp_eager_bcopy_single_progress()`), or I don't see other usages?

File: src/ucp/proto/proto_multi.inl (hunk 1 above)
Comment (yosefe) [reply src/ucp/proto/proto_multi.inl]: added comment

File: src/ucp/proto/proto_single.inl (hunk 2 above)
Comment (yosefe) [reply src/ucp/proto/proto_single.inl:27]: max_packed_size could be compile-time constant sometimes

File: src/ucp/proto/proto_single.inl (hunk 3 above)
Comment (yosefe) [reply src/ucp/proto/proto_single.inl]: added comment

File: src/ucp/tag/eager_single.c (hunk 4 above)
Comment (yosefe) [reply src/ucp/tag/eager_single.c]: fixed

File: src/ucp/proto/proto_single.inl (hunk 5 above)
Comment (yosefe) [reply src/ucp/proto/proto_single.inl:62]: it's in code which will be in later PR

**Review by dmitrygx — APPROVED** [review]

File: src/ucp/tag/eager_multi.c [hunk 6]
```diff
@@ -162,15 +162,57 @@ ucp_proto_eager_zcopy_multi_init(const ucp_proto_init_params_t *init_params)
...
+    return uct_ep_am_zcopy(req->send.ep->uct_eps[lpriv->super.lane], am_id, &hdr,
+                           hdr_size, &iov, 1, 0, &req->send.state.uct_comp);
+}
+
+static ucs_status_t ucp_proto_eager_zcopy_multi_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+
+    if (!(req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED)) {
+        ucp_eager_multi_proto_request_init(req);
```
Comment (brminich) [inline src/ucp/tag/eager_multi.c:203]: maybe call this init function in `ucp_proto_multi_zcopy_progress` to have all initializations in one place?

**Review by brminich — APPROVED** [review]

**Pull request merged**